docs/rfc/rfc-105-event-observer.md

```diff
@@ -0,0 +1,240 @@
+information from other reactors/services: the [`Switch`] allows developers to
+look up reactors, at runtime, and access methods directly on those reactors.
+This is again an inappropriate pattern because all lookups are hard-coded, and
+reactors/services are not dynamically created/destroyed at runtime.
+
+This suggests that a different approach is necessary for cross-component
+interaction - ideally one which provides more robust compile-time guarantees
+than the current ones.
+
+## Discussion
+
+A more type-safe, understandable Observer pattern is proposed here than what
+currently exists in CometBFT. For example, in the consensus state there are many
+places where events are published via the event bus, e.g.:
+
+- <https://github.com/cometbft/cometbft/blob/091a1f312e5f2f4b183fab1d57d729a6c478ff1f/consensus/state.go#L743>
+- <https://github.com/cometbft/cometbft/blob/091a1f312e5f2f4b183fab1d57d729a6c478ff1f/consensus/state.go#L964>
+- <https://github.com/cometbft/cometbft/blob/091a1f312e5f2f4b183fab1d57d729a6c478ff1f/consensus/state.go#L971>
+- etc.
+
+All of these event publishing methods, not only for consensus state but for
+other types of event publishers, are defined on this central `EventBus` type,
+which seems to signal that its functionality should be decentralized.
+
+### Strongly Typed Event Observer
+
+A simple alternative pattern here would be to define an **event observer**
+interface for every major component of the system capable of producing events.
+For consensus state, this may end up looking something like:
+
+```golang
+package consensus
+
+// StateObserver is specific to the consensus.State struct, and all of its
+// methods are called from within consensus.State instead of using an "event
+// bus". This allows for greater compile-time guarantees through doing away with
+// the generic pub/sub mechanism in the event bus.
+//
+// Note how all methods are infallible (i.e. they do not return any errors).
+// This is functionally equivalent to the fire-and-forget pattern implemented by
+// the event bus.
+//
+// Also note how method names are prefixed by the name of the relevant producer
+// of events (in this case "ConsensusState", corresponding to the
+// consensus.State struct). This is intentional to allow composition of
+// observers of multiple different components without function names clashing.
+//
+// Finally, given that this is just straightforward Go, it is up to either the
+// caller or the callee to decide how to handle the concurrency of certain
+// events. The event bus approach, by contrast, is always concurrent and relies
+// on Go channels, which could end up filling up and causing back-pressure into
+// the caller (already observed in slow WebSocket subscribers).
+type StateObserver interface {
+    ConsensusStateNewRoundStep(ev EventDataRoundState)
+    ConsensusStateTimeoutPropose(ev EventDataRoundState)
+    ConsensusStateTimeoutWait(ev EventDataRoundState)
+    // ...
+}
```

Comment:
If we call it `ConsensusStateObserver`, and remove the prefix from the methods, would it cause name clash with other `XxxObserver`, `YyyObersver`, interfaces?

Actually, since the interface is part of the Consensus package, we wouldn't even need the `Consensus...`, since every time it's used outside the package, we'd need to prefix with the package name.

Reply:
This was a conscious decision on my part to name the functions as such - perhaps I should elaborate on that in the RFC. This allows for observer composition. For instance, if we have two observer interfaces, the `StateObserver` one and another interface:

```go
type OtherObserver interface {
    OtherTimeoutWait(params OtherTimeoutParams)
}
```

We could imagine an observer that implements both interfaces because it's interested in both events:

```go
type CombinedObserver interface {
    StateObserver
    OtherObserver
}
```

If we strip the `ConsensusState` and `Other` prefixes from the function names, the two would clash and there would be no way for anyone who doesn't control the interface definitions (e.g. external consensus engine developers) to implement a single observer for both sets of events.

Therefore we need to make sure that all observers' function names are globally unique throughout our codebase.